File: scala/3.scala
import scala.collection.mutable
object Solution {
    def lengthOfLongestSubstring(s: String): Int = {
      var l = 0
      var r = 0
      var set = mutable.Set[Char]()
      var res = 0
      while ( r < s.length() ) {
        if (set.contains(s(r))) {
          while ( s(l) != s(r) ) { 
              set.remove(s(l))
              l += 1
          }
          l += 1
        } else {
          set.add(s(r))
        }
        r += 1
        res = res.max(r-l)
      }
      res
    }
}

File: go/3.go
func lengthOfLongestSubstring(s string) int {
	indexMap := map[byte]int{}
	l, r := 0, 0
	res := 0
	for r < len(s) {
		if i, ok := indexMap[s[r]]; ok {
			for k, v := range indexMap {
				if v < i {
					delete(indexMap, k)
				}
			}
			l = i + 1
		}
		indexMap[s[r]] = r

		if r-l+1 > res {
			res = r - l + 1
		}
		r++
	}
	return res
}